mscore/resourceManager.cpp

```diff
@@ -264,6 +414,343 @@ bool ResourceManager::verifyLanguageFile(QString filename, QString hash)
+      Q_ASSERT(desc.update != nullptr);
+      if (desc.update->source == UNKNOWN) {
+            install->setText("Failed Try again.");
+            install->setEnabled(true);
+            return;
+            }
+      install->setText(tr("Downloading..."));
+      QString localPath = downloadPluginPackage(*desc.update, page_url);
+      if (installPluginPackage(localPath, *desc.update)) {
+            PluginPackageDescription* p_update = desc.update;
+            desc = *desc.update;
+            delete(p_update);
+            desc.update = nullptr;
+            refreshPluginButton(install->property("row").toInt(), true);
+            writePluginPackages();
+            }
+      }
+
+void ResourceManager::checkPluginUpdate(QPushButton* install)
+      {
+      const QString& page_url = install->property("page_url").toString();
+      if (pluginDescriptionMap.contains(page_url)) {
+            auto& desc = pluginDescriptionMap[page_url];
+            PluginPackageDescription* desc_tmp = new PluginPackageDescription(desc);
+            bool should_update = analyzePluginPage("https://musescore.org" + page_url, *desc_tmp);
+            if (should_update)
+                  desc.update = desc_tmp;
+            else
+                  delete desc_tmp;
+            refreshPluginButton(install->property("row").toInt(), !should_update);
+            }
+      }
+
+void ResourceManager::scanPluginUpdate()
+      {
+      for (int row = 0; row < pluginsTable->rowCount(); row++) {
+            QPushButton* install = static_cast<QPushButton*>(pluginsTable->indexWidget(pluginsTable->model()->index(row, 3)));
+            const QString& page_url = install->property("page_url").toString();
+            // if installed
+            if (pluginDescriptionMap.contains(page_url))
+                  checkPluginUpdate(install);
+            }
+      }
+
+//---------------------------------------------------------
+//   installPluginPackage
+//---------------------------------------------------------
+
+bool ResourceManager::installPluginPackage(QString& download_pkg, PluginPackageDescription& desc)
+      {
+      QFileInfo f_pkg(download_pkg);
+      QString suffix = f_pkg.suffix().toLower();
+      QString destination_prefix = dataPath + "/plugins" + "/" + f_pkg.completeBaseName();
+      QDir().mkdir(destination_prefix);
+      desc.dir = destination_prefix;
+      if (suffix == "zip") {
+            MQZipReader zipFile(download_pkg);
+            QVector<MQZipReader::FileInfo> allFiles = zipFile.fileInfoList();
+            if (allFiles.size() == 0)
+                  return false;
+            // If zip contains multiple files in root, or a single qml, create a
+            // root folder in plugin dir for them.
+            // If zip contains a single directory, don't use that directory's name
+            bool has_no_dir = true;
+            std::set<QString> dirs;
+            foreach(MQZipReader::FileInfo fi, allFiles) {
```

Comment:
Please `for (MQZipReader::FileInfo fi : allFiles) {`

Reply:
Resolved in 77701599820befd5e5381af2075d63e6b5b3dd85.